dirty_cat/super_vectorizer.py

```diff
@@ -0,0 +1,341 @@
+
+    def _auto_cast_array(self, X):
+        """
+        Takes an array and tries to convert its columns to the best possible
+        data type.
+
+        Parameters
+        ----------
+        X: array-like
+            Input data.
+
+        Returns
+        -------
+        array
+            The same array, with its columns casted to the best possible
+            data type.
+
+        Raises
+        ------
+        RuntimeError
+            If no transformers could be constructed,
+            usually because transformers passed do not match any column.
+            To fix the issue, try passing the least amount of None as encoders.
+
+        """
+
+        def cast_astype(col):
+            # First, try to convert the column to floats
+            try:
+                return col.astype(float)
+            except ValueError:
+                # Couldn't cast
+                pass
+
+            # Next, to integers
+            try:
+                return col.astype(int)
+            except ValueError:
+                pass
+
+            # Finally, to strings
+            # We are not using try-except because this should work no matter the data.
+            # (to be confirmed).
+            return col.astype(str)
+
+        def cast_series(sr: pd.Series) -> pd.Series:
+            if _has_missing_values(sr):
+                raise ValueError(_ERR_MSG_FOUND_NAN)
+
+            return cast_astype(sr)
+
+        def cast_dataframe(df: pd.DataFrame) -> pd.DataFrame:
+            return df.convert_dtypes()
+
+        def cast_array(arr: np.array) -> np.array:
+
+            def cast_column(col: np.array):
+                """
+                Method to call on each column,
+                which will try to cast said column to the best possible type.
+                """
+
+                # Check if the array contains NaN or INF
+                if _has_missing_values(col):
+                    raise ValueError(_ERR_MSG_FOUND_NAN)
+                if any(col.isinf()):
+                    raise ValueError(_ERR_MSG_FOUND_INF)
+
+                return cast_astype(col)
+
+            try:
+                # nD array
+                arr = np.apply_along_axis(func1d=cast_column, axis=1, arr=arr)
+            except np.AxisError:
+                # 1D array
+                arr = cast_column(arr)
+
+            return arr
+
+        if isinstance(X, pd.Series):
+            return cast_series(X)
+        elif isinstance(X, pd.DataFrame):
+            return cast_dataframe(X)
+        elif isinstance(X, np.ndarray):
+            return cast_array(X)
+
+    def _transform(self, X) -> pd.DataFrame:
+        # Convert to pandas DataFrame if not already.
+        if not isinstance(X, pd.DataFrame):
+            X = pd.DataFrame(X)
+
+        # Detect if the array contains missing values.
+        if _has_missing_values(X):
+            if self.handle_missing == '':
+                X = _replace_missing(X)
```

Comment:
Actually, ideally I think that we would not handle missing values in the supervectorizer, as often downstream (transformers or even the final learner) can do this better.